### PR TITLE
feat: update to TypeScript 4.7-rc

### LINKED ISSUE #4800: TypeScript 4.7 Support
https://devblogs.microsoft.com/typescript/announcing-typescript-4-7-rc/

This issue is just to track all of the new features and their implementation state in this project.
As with all releases, we will not necessarily to support all features until closer to the full release when everything the features are stabilised.

*Please be patient.*

# 🔲  New Extensions

We'll want to add `.cjs`, `.mjs`, `.mts`, and `.cts` (as well as `.d.cts` and `.d.mts`) to our parser tooling so that we can correctly detect file types and default supported extensions.

Some code pointers to start off:
[[1]](https://github.com/typescript-eslint/typescript-eslint/blob/4c428c5840f113385dc3c91a06a781d2c1f7642d/packages/typescript-estree/src/create-program/shared.ts#L75-L102) [[2]](https://github.com/typescript-eslint/typescript-eslint/blob/4c428c5840f113385dc3c91a06a781d2c1f7642d/packages/typescript-estree/src/create-program/shared.ts#L104-L109) [[3]](https://github.com/typescript-eslint/typescript-eslint/blob/4c428c5840f113385dc3c91a06a781d2c1f7642d/packages/typescript-estree/src/create-program/createProjectProgram.ts#L10) [[4]](https://github.com/typescript-eslint/typescript-eslint/blob/4c428c5840f113385dc3c91a06a781d2c1f7642d/packages/eslint-plugin/src/configs/eslint-recommended.ts#L9) [[5]](https://github.com/typescript-eslint/typescript-eslint/blob/4c428c5840f113385dc3c91a06a781d2c1f7642d/packages/eslint-plugin/src/util/misc.ts#L11-L13)

Essentially just search for `.ts` and go from there.

# ✅ Instantiation Expressions (https://github.com/typescript-eslint/typescript-eslint/pull/4938)

This will require AST changes.

Babel: https://github.com/babel/babel/pull/14457

```ts
// extend Expression
type Expression = Expression | TSExpressionWithTypeArguments;

// Add new node
interface TSExpressionWithTypeArguments {
  type: "TSExpressionWithTypeArguments";
  expression: Expression;
  typeParameters: TSTypeParameterInstantiation;
}
```

# ✅ `extends` Constraints on `infer` Type Variables (https://github.com/typescript-eslint/typescript-eslint/pull/4830)

This is already supported and will not require AST changes

# ✅ Optional Variance Annotations for Type Parameters (https://github.com/typescript-eslint/typescript-eslint/pull/4831)

This will require AST changes.

Babel: https://github.com/babel/babel/pull/14359

```ts
// extend TSTypeParameter
interface TSTypeParameter {
  in?: boolean;
  out?: boolean;
}
```

# ✅ `lib.d.ts` Updates (https://github.com/typescript-eslint/typescript-eslint/pull/4829)

We will need to regenerate our types within `scope-manager`.

---

# Other changes with no impact to us

- ECMAScript Module Support in Node.js
- `type` in `package.json`
- CommonJS Interop
- `package.json` Exports, Imports, and Self-Referencing
  - Though with the next major we'll probably want to explore this to lock down our APIs and prevent deep `dist` imports whilst also providing nicely structured deep imports (i.e. map `utils/dist/foo` => `utils/foo`)
- Control over Module Detection
- Control-Flow Analysis for Computed Properties
  - This will just make all of our rules more accurate for free!
- Improved Function Inference in Objects and Methods
- Resolution Customization with `moduleSuffixes`
- resolution-mode
- Groups-Aware Organize Imports
- Object Method Snippet Completions
- Type Parameters No Longer Assignable to {} in strictNullChecks
- `readFile` Method is No Longer Optional on `LanguageServiceHost`
- `readonly` Tuples Have a `readonly length` Property

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## PR Checklist

-   [x] Addresses an existing open issue: fixes #4800 
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

note: ~~`api-extractor` does not working on this PR, so `d.ts` rollup for `ast-spec` does not working too. So all CI is failing...~~ https://github.com/microsoft/rushstack/issues/3349 has been fixed!

- Updates to TypeScript 4.7
- `generate:lib`
- Avoid some type errors.
- ~~Avoid some syntax errors.~~
  - https://github.com/microsoft/TypeScript/issues/48733

### CODE REVIEW

File: packages/ast-spec/package.json [hunk 1]
```diff
@@ -29,7 +29,7 @@
   "main": "dist/index.js",
   "types": "dist/index.d.ts",
   "scripts": {
-    "build": "tsc -b tsconfig.build.json && api-extractor run --local",
+    "build": "tsc -b tsconfig.build.json",
```
Comment (sosukesuzuki) [inline packages/ast-spec/package.json:32]: `api-extractor` seems to be not working with 4.7-beta (ref: https://github.com/microsoft/rushstack/issues/3349). So I've removed `api-extractor` step temporary.

File: packages/utils/src/ast-utils/helpers.ts [hunk 2]
```diff
@@ -26,9 +26,7 @@ export const isNodeOfTypeWithConditions = <
 ): ((
   node: TSESTree.Node | null | undefined,
 ) => node is TSESTree.Node & { type: NodeType } & Conditions) => {
-  const entries = Object.entries(conditions) as ObjectEntries<
-    TSESTree.Node & { type: NodeType }
-  >;
+  const entries = Object.entries(conditions) as ObjectEntries<TSESTree.Node>;
```
Comment (sosukesuzuki) [inline packages/utils/src/ast-utils/helpers.ts:29]: I have no idea to fix type errors without removing `& { type: NodeType }`...

File: packages/utils/src/ast-utils/helpers.ts [hunk 3]
```diff
@@ -46,9 +44,7 @@ export const isTokenOfTypeWithConditions = <
 ): ((
   token: TSESTree.Token | null | undefined,
 ) => token is TSESTree.Token & { type: TokenType } & Conditions) => {
-  const entries = Object.entries(conditions) as ObjectEntries<
-    TSESTree.Token & { type: TokenType }
-  >;
+  const entries = Object.entries(conditions) as ObjectEntries<TSESTree.Token>;
```
Comment (sosukesuzuki) [inline packages/utils/src/ast-utils/helpers.ts:47]: ditto

File: packages/ast-spec/package.json (hunk 1 above)
Comment (sosukesuzuki) [reply packages/ast-spec/package.json]: fixed by https://github.com/microsoft/rushstack/pull/3394

**Comment by bradzacher** [comment]
We can look to merging this now as the RC has been released.

File: packages/utils/src/ast-utils/helpers.ts (hunk 2 above)
Comment (bradzacher) [reply packages/utils/src/ast-utils/helpers.ts:29]: This looks fine to me!

**Review by bradzacher — APPROVED** [review]
LGTM - thanks for this!

**Pull request merged**